Assignment2Q2.java

```diff
@@ -0,0 +1,129 @@
+
+public class Assignment2Q2 {
+	public static void main(String[] args) {
+		// construct the tree
+		BinaryTreeNode<Integer> n1 = new BinaryTreeNode<>(1, null, null);
+		BinaryTreeNode<Integer> n2 = new BinaryTreeNode<>(5, null, null);
+		BinaryTreeNode<Integer> n3 = new BinaryTreeNode<>(3, n1, n2);
+		BinaryTreeNode<Integer> n4 = new BinaryTreeNode<>(14, null, null);
+		BinaryTreeNode<Integer> n5 = new BinaryTreeNode<>(9, n3, n4);
+		BinaryTreeNode<Integer> n6 = new BinaryTreeNode<>(19, null, null);
+		BinaryTreeNode<Integer> n7 = new BinaryTreeNode<>(18, null, n6);
+		BinaryTreeNode<Integer> n8 = new BinaryTreeNode<>(16, n5, n7);
+		BinaryTreeNode<Integer> n9 = new BinaryTreeNode<>(186, null, null);
+
+		System.out.println(findCommonAncestors(n9, n1, n8).getData());
+	}
+
+	/**
+	 * Finds lowest common ancestor of node1 and node2
+	 * 
+	 * @param node1
+	 *            first node
+	 * @param node2
+	 *            second node
+	 * @param root
+	 *            root of the tree
+	 * @return lowest common ancestor of node1 and node2
+	 * @throws IllegalArgumentException
+	 *             if one or both of the nodes are not present in the tree
+	 */
+	public static <T> BinaryTreeNode<T> findCommonAncestors(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2,
+			BinaryTreeNode<T> root) throws IllegalArgumentException {
+		if (node1 == null || node2 == null || root == null) {
+			throw new IllegalArgumentException();
+		}
+
+		BinaryTreeNode<T> node = findCommonAncestorsHelper(node1, node2, root);
+
+		if (node == null) {
+			throw new IllegalArgumentException();
+		}
+
+		// if node1 is the common ancestor, check if node2 is in the tree which
+		// has root node1. Also handles the case if node1 and node2 are the
+		// same.
+		if (node.getData().equals(node1.getData())) {
+			if (node1.getData().equals(node2.getData()) || contains(node2, node)) {

```

Comment:
The node.getData().equals(node2.getData()) pattern appears quite a lot across the code. You could consider adding something like containsData(T data) method to node and use it instead.